core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -299,6 +300,54 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     }
   }
 
+  override def readEventLogs(zipStream: ZipInputStream): Unit = {
+    val fs = FileSystem.get(hadoopConf)
+    val tmpPath = new Path(logDir, UUID.randomUUID().toString + TEMP_IMPORT_LOG_DIR_SUFFIX)
+    var zipEntry = zipStream.getNextEntry
+
+    try {
+      // Create tmp dir to store unzipped files temporarily
+      // Reason to put unzipped files in this temporary file is to avoid contention state when
+      // checkForLogs is occurred during unzipping.
+      fs.mkdirs(tmpPath)
+
+      while (zipEntry != null) {
+        logInfo(s"Unzipping ${zipEntry.getName}")
+        val path = new Path(tmpPath, zipEntry.getName)
+        if (zipEntry.isDirectory) {
+          // this could possibly be legacy history log
+          fs.mkdirs(path)
+        } else {
+          var out: OutputStream = null
+          Utils.tryWithSafeFinally {
+            out = fs.create(path, true, 1 * 1024 * 1024)
+            Utils.copyStream(zipStream, out)
+          } {
+            if (out != null) {
```

Comment:
Hadoop`s `IOUtils.closeStream` can do the close here, including null check & log problems @ debug